### PR TITLE
[BART] Bart Fixes

### PR DESCRIPTION
**Pull request by klshuster**
**Patch description**
BART was ... having some issues. There's an odd thing about BART where it expects `EOS BOS` tokens prior to generation. This makes training very screwy.

The solution is to prepend `EOS BOS` to the label during training, but then *ignore* the first scored output from BART when computing perplexity. Please, I am not sure I can explain why, so don't ask.

**Testing steps**
Updated `test_bart.py` to test the following desired scenarios:

1. BART, out of the box, generates correctly (as an autoencoder, this means it just copies the input)
2. BART, out of the box, scores perfectly on the default integration test (whose labels are repeats of the text)
3. BART, fine-tuned on a new, different task, gets near-1 ppl (very quickly)
4. BART, fine-tuned on a new, different task, generates correctly.

**Logs**
```
$ python -m pytest tests/nightly/gpu/test_bart.py
================================================================================ test session starts ================================================================================
platform linux -- Python 3.6.9, pytest-5.3.2, py-1.8.1, pluggy-0.13.1
rootdir: /private/home/kshuster/ParlAI, inifile: pytest.ini
plugins: requests-mock-1.7.0
collected 3 items

tests/nightly/gpu/test_bart.py ...                                                                                                                                            [100%]

============================================================================= slowest 10 test durations =============================================================================
96.27s call     tests/nightly/gpu/test_bart.py::TestBartModel::test_bart_ft

(0.00 durations hidden.  Use -vv to show these durations.)
==================================================================== 3 passed, 39 warnings in 163.47s (0:02:43) =====================================================================
```

### CODE REVIEW

**Comment by emilydinan** [comment]
Screen Shot 2020-08-28 at 4 36 59 PM

**Comment by klshuster** [comment]
I don't understand why lint is failing...

```
$ bash autoformat.sh
All done! ✨ 🍰 ✨
4 files left unchanged.
```

File: parlai/agents/bart/modules.py [hunk 1]
```diff
@@ -27,26 +26,37 @@ def output(self, tensor: torch.Tensor) -> torch.Tensor:
...
-        seqlen = ys.size(1)
-        inputs = ys.narrow(1, 0, seqlen - 1)
-        inputs = torch.cat(
-            [
-                torch.LongTensor([self.END_IDX]).detach().expand(bsz, 1).to(inputs),
-                inputs,
-            ],
-            1,
+        tens = (
+            torch.LongTensor([self.END_IDX, self.START_IDX])
```
Comment (EricMichaelSmith) [inline parlai/agents/bart/modules.py:44]: Hmm we could abstract out this intro sequence (`[self.END_IDX, self.START_IDX]`), which shows up at least twice

File: parlai/agents/bart/modules.py [hunk 2]
```diff
@@ -27,26 +26,37 @@ def output(self, tensor: torch.Tensor) -> torch.Tensor:
...
+        incremental_state: Dict[str, Any],
+        inds: Union[List[int], torch.LongTensor],
+    ) -> Optional[Dict[str, Any]]:
+        """
+        Incremental state is weird to handle when we seed decoder with two inputs
+        initially.
+
+        We leave as a future exercise.
+        """
+        return None
```
Comment (EricMichaelSmith) [inline parlai/agents/bart/modules.py:62]: Hmm so I imagine generation with BERT will be a lot slower (but, er, correct) now?

Comment (klshuster) [reply parlai/agents/bart/modules.py:62]: slower yes; a lot slower? maybe; probably not too significantly. if it becomes unwieldy in the future we can address then

File: parlai/agents/bart/modules.py (hunk 1 above)
Comment (klshuster) [reply parlai/agents/bart/modules.py:44]: I'm not sure we can, it shows up in separate classes (`BartAgent` vs. `BartModel`) and is dependent on the set attrs for `END_IDX` and `START_IDX` which we do not know beforehand

File: tests/nightly/gpu/test_bart.py [hunk 3]
```diff
@@ -4,35 +4,76 @@
...
-                dict_file='zoo:bart/bart_large/model.dict',
-                optimizer='sgd',
-                learningrate=1,
-                batchsize=4,
-                num_epochs=1,
-                short_final_eval=True,
-                validation_max_exs=12,
-            )
+        """
+        Test out-of-the-box BART on repeat task.
```
Comment (stephenroller) [inline tests/nightly/gpu/test_bart.py:24]: nice

**Comment by stephenroller** [comment]
You can ignore the black complaints.

File: tests/nightly/gpu/test_bart.py [hunk 4]
```diff
@@ -4,35 +4,76 @@
...
         )
-        self.assertLessEqual(valid['ppl'], 11.0)
-        self.assertLessEqual(test['ppl'], 11.0)
+        self.assertAlmostEqual(valid['ppl'].value(), 1.0, places=1)
+
+    def test_bart_gen(self):
+        """
+        Test out-of-the-box BART generation.
+        """
+        opt = ParlaiParser(True, True).parse_args(['--model', 'bart'])
```
Comment (stephenroller) [inline tests/nightly/gpu/test_bart.py:35]: you can do `ParlaiParser(True, True).parse_kwargs(model='bart')` if you like that better
